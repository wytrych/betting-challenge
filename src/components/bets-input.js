import React, { Component } from 'react'
import { parseBetsList } from '../utils/parsers'
import { markErroredLines, stripErrorMarks } from '../utils/utils'

export class BetsInput extends Component {
    constructor (props) {
        super(props)

        this.state = {
            input: '',
            error: '',
            success: '',
        }

        this.parseForm = this.parseForm.bind(this)
        this.update = this.update.bind(this)
    }

    parseForm (e) {
        e.preventDefault()
        this.setState((prevState) => {
            let error = ''
            let success = ''
            let input = stripErrorMarks(prevState.input).replace(/ /g, '\n')

            try {
                const bets = parseBetsList(input.split('\n'))
                this.props.onChange(bets)
                success = 'Success!'
            } catch (e) {
                error = e.message
                input = markErroredLines(input, e.erroredLines)
            }

            return {
                error,
                success,
                input,
            }
        })
    }

    update (e) {
        this.setState({input: e.target.value})
    }

    render () {
        return (
            <form onSubmit={this.parseForm} className={this.state.error ? 'has-error' : ''}>
                <div className='form-group'>
                    <label className='control-label' htmlFor='bets'>Bets</label>
                    <textarea id='bets' className='form-control' rows='10' value={this.state.input} onChange={this.update}></textarea>
                </div>
                <button className='btn btn-primary' disabled={!this.state.input}>Submit bets</button>

                <div className='help-block'>
                    {this.state.error.split('\n').map((errorLine, i) => (
                        <div key={i}>{errorLine}</div>
                    ))}
                    {this.state.success}
                </div>
            </form>
        )
    }
}
