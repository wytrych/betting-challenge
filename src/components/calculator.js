import React, { Component } from 'react'
import { addSpareLines, limit } from '../utils/utils'
import { DEFAULT_COMISSIONS } from '../calculators/calculators'
import { calculateDividends } from '../calculators/calculator-dispatcher'

export class Calculator extends Component {
    constructor (props) {
        super(props)
        this.state = {
            comission: DEFAULT_COMISSIONS[props.type] * 100,
        }

        this.counter = this.counter.bind(this)
        this.updateComission = this.updateComission.bind(this)
    }

    updateComission (e) {
        this.setState({
            comission: limit(e.target.value, 0, 100)
        })
    }

    currencySign (dividend) {
        return isFinite(parseFloat(dividend)) ? '$ ' : ''
    }

    counter (i) {
        return this.props.showKey ? `${i + 1}: ` : ''
    }

    render () {
        const dividends = calculateDividends(this.props.type, this.props.bets, this.props.result, this.state.comission / 100)
        const paddedDividends = addSpareLines(dividends)
        return (
            <div className='dividend col-md-3 col-xs-6'>
                <h3>{this.props.children}</h3>
                <div className='well'>
                    {paddedDividends.map((dividend, i) => (
                        <div key={i}><h4><strong>{this.counter(i)}{this.currencySign(dividend)}{dividend}</strong></h4></div>
                    ))}
                </div>
                <div className='form-group'>
                    <label>Comission rate (%):
                        <input value={this.state.comission} onChange={this.updateComission} type='range' min='0' max='100' />
                        <input className='form-control' value={this.state.comission} onChange={this.updateComission} type='number' min='0' max='100' />
                    </label>
                </div>
            </div>
        )
    }
}
