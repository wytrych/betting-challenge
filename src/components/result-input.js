import React, { Component } from 'react'
import { parseResult } from '../utils/parsers'

export class ResultInput extends Component {
    constructor (props) {
        super(props)
        this.state = {
            result: '',
            error: '',
            success: '',
        }

        this.addResult = this.addResult.bind(this)
        this.updateResult = this.updateResult.bind(this)
    }

    updateResult (e) {
        this.setState({result: e.target.value })
    }

    addResult (e) {
        e.preventDefault()
        let error = ''
        let success = ''

        try {
            const result = parseResult(this.state.result)
            this.props.onChange(result)
            success = 'Success'
        } catch (e) {
            error = e.message
        }

        this.setState({
            error,
            success,
        })
    }

    render () {
        return (
            <form className={'form-inline ' + (this.state.error ? 'has-error' : '')} onSubmit={this.addResult}>
                <div className='form-group'>
                    <label className='control-label' htmlFor='result'>Result:</label>
                    <input id='result' className='form-control' type='text' value={this.state.result} onChange={this.updateResult} />
                </div>
                <button className='btn btn-info' disabled={!this.state.result}>Update result</button>
                <div className='help-block'>{this.state.error}{this.state.success}</div>
            </form>
        )
    }
}
