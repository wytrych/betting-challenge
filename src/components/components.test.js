/* eslint-env jest */

import React from 'react'
import renderer from 'react-test-renderer'
import { BetsInput } from './bets-input'
import { Calculator } from './calculator'
import { ResultInput } from './result-input'

describe('bets-input', () => {

    it('should render', () => {
        const tree = renderer.create(<BetsInput />)
        expect(tree).toMatchSnapshot()
    })

})

describe('Calculator', () => {

    it('should render', () => {
        const bets = {W: [] }
        const tree = renderer.create(<Calculator bets={bets} type='W' />)
        expect(tree).toMatchSnapshot()
    })

})

describe('ResultInput', () => {

    it('should render', () => {
        const tree = renderer.create(<ResultInput />)
        expect(tree).toMatchSnapshot()
    })

})
