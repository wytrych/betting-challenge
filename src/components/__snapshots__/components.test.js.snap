exports[`Calculator should render 1`] = `
<div
  className="dividend col-md-3 col-xs-6">
  <h3 />
  <div
    className="well">
    <div>
      <h4>
        <strong>
          
          
           
        </strong>
      </h4>
    </div>
    <div>
      <h4>
        <strong>
          
          
          No winning bets
        </strong>
      </h4>
    </div>
    <div>
      <h4>
        <strong>
          
          
           
        </strong>
      </h4>
    </div>
  </div>
  <div
    className="form-group">
    <label>
      Comission rate (%):
      <input
        max="100"
        min="0"
        onChange={[Function]}
        type="range"
        value={15} />
      <input
        className="form-control"
        max="100"
        min="0"
        onChange={[Function]}
        type="number"
        value={15} />
    </label>
  </div>
</div>
`;

exports[`ResultInput should render 1`] = `
<form
  className="form-inline "
  onSubmit={[Function]}>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="result">
      Result:
    </label>
    <input
      className="form-control"
      id="result"
      onChange={[Function]}
      type="text"
      value="" />
  </div>
  <button
    className="btn btn-info"
    disabled={true}>
    Update result
  </button>
  <div
    className="help-block">
    
    
  </div>
</form>
`;

exports[`bets-input should render 1`] = `
<form
  className=""
  onSubmit={[Function]}>
  <div
    className="form-group">
    <label
      className="control-label"
      htmlFor="bets">
      Bets
    </label>
    <textarea
      className="form-control"
      id="bets"
      onChange={[Function]}
      rows="10"
      value="" />
  </div>
  <button
    className="btn btn-primary"
    disabled={true}>
    Submit bets
  </button>
  <div
    className="help-block">
    <div>
      
    </div>
    
  </div>
</form>
`;
